File: src/lib/religious-voices/rate-limit.ts
// In-memory per-IP token bucket. Ephemeral; resets on cold start. Fine
// for a portfolio site — production-grade abuse prevention would back
// this with Redis or DynamoDB.

const CAPACITY = 20;
const REFILL_PER_SEC = CAPACITY / (60 * 60); // 20 tokens / hour

interface Bucket {
  tokens: number;
  lastRefill: number;
}

const buckets = new Map<string, Bucket>();

export function checkRateLimit(ip: string): { allowed: boolean; retryAfter?: number } {
  const now = Date.now() / 1000;
  let b = buckets.get(ip);
  if (!b) {
    b = { tokens: CAPACITY, lastRefill: now };
    buckets.set(ip, b);
  } else {
    b.tokens = Math.min(CAPACITY, b.tokens + (now - b.lastRefill) * REFILL_PER_SEC);
    b.lastRefill = now;
  }
  if (b.tokens >= 1) {
    b.tokens -= 1;
    return { allowed: true };
  }
  const retryAfter = Math.max(1, Math.ceil((1 - b.tokens) / REFILL_PER_SEC));
  return { allowed: false, retryAfter };
}

File: projects/religious-voices/server/rate_limit.py
"""In-memory per-IP token bucket — Python port of the TS rate-limit.ts.

Ephemeral process-local state; resets on restart. Fine for a portfolio
site. Production-grade abuse prevention would back this with Redis.
"""

from __future__ import annotations

import time
from dataclasses import dataclass

CAPACITY = 20  # max messages per window
REFILL_PER_SEC = CAPACITY / (60 * 60)  # 20 tokens per hour


@dataclass
class _Bucket:
    tokens: float
    last_refill: float


_buckets: dict[str, _Bucket] = {}


def check_rate_limit(ip: str) -> tuple[bool, int | None]:
    """Returns (allowed, retry_after_seconds)."""
    now = time.time()
    b = _buckets.get(ip)
    if b is None:
        b = _Bucket(tokens=CAPACITY, last_refill=now)
        _buckets[ip] = b
    else:
        elapsed = now - b.last_refill
        b.tokens = min(CAPACITY, b.tokens + elapsed * REFILL_PER_SEC)
        b.last_refill = now
    if b.tokens >= 1:
        b.tokens -= 1
        return True, None
    secs = max(1, int((1 - b.tokens) / REFILL_PER_SEC))
    return False, secs
